### PR TITLE
GH-40841: [Docs][C++][Python] Add initial documentation for RecordBatch::Tensor conversion

### PR DESCRIPTION
**Pull request by AlenkaF**
### Rationale for this change

The work on the conversion from `Table`/`RecordBatch` to `Tensor` is progressing and we have to make sure to add information to the documentation. 

### What changes are included in this PR?

I propose to add

- new page (`converting_recordbatch_to_tensor.rst`) in the `cpp/examples` section,
- added section (Conversion of RecordBatch do Tensor) in the `docs/source/python/data.rst`

the content above would be updated as the features are added in the future (row-major conversion, `Table::ToTensor`, DLPack support for `Tensor` class, etc.)

### Are these changes tested?

It will be tested with the crossbow preview-docs job.

### Are there any user-facing changes?

No, just documentation.
* GitHub Issue: #40841

### CODE REVIEW

**Comment by AlenkaF** [comment]
@github-actions crossbow submit preview-docs

**Comment by AlenkaF** [comment]
@github-actions crossbow submit preview-docs

**Comment by AlenkaF** [comment]
Links to the content of this PR in the preview build:

- http://crossbow.voltrondata.com/pr_docs/40842/cpp/examples/converting_recordbatch_to_tensor.html
- http://crossbow.voltrondata.com/pr_docs/40842/python/data.html#conversion-of-recordbatch-do-tensor

File: docs/source/cpp/examples/converting_recordbatch_to_tensor.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+
+Conversion of ``RecordBatch`` to ``Tensor`` instances
+=====================================================
+
+Arrow provides a method to convert ``RecordBatch`` objects to ``Tensors``
```
Comment (jorisvandenbossche) [inline docs/source/cpp/examples/converting_recordbatch_to_tensor.rst:24]:
```suggestion
Arrow provides a method to convert ``RecordBatch`` objects to a ``Tensor``
```

File: docs/source/cpp/examples/converting_recordbatch_to_tensor.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+   std::shared_ptr<RecordBatch> batch;
+
+   ASSERT_OK_AND_ASSIGN(auto tensor, batch->ToTensor());
+   ASSERT_OK(tensor->Validate());
+
+The conversion supports signed and unsigned integer types plus float types,
+all widths included. In case the ``RecordBatch`` has null values the conversion
+succeeds if ``null_to_nan`` parameter is set to ``true``. In this case all
+types will be promoted to float-point data type.
```
Comment (jorisvandenbossche) [inline docs/source/cpp/examples/converting_recordbatch_to_tensor.rst:37]:
```suggestion
types will be promoted to a floating-point data type.
```

File: docs/source/python/data.rst [hunk 3]
```diff
@@ -560,3 +560,59 @@ schema without having to get any of the batches.::
...
+Each array of the ``RecordBatch`` has it's own contiguous memory that is not necessarily
+adjacent to other arrays. A different memory structure that is used in machine learning
+libraries is a two dimensional array (also called a 2-dim tensor or a matrix) which takes
+only one contiguous block of memory.
+
+For this reason there is a function ``pyarrow.RecordBatch.to_tensor()`` available
+to efficiently convert tabular columnar data into a matrix.
+
+Data types supported in this conversion are unsigned, signed integer and float
+types of all widths. Currently only column-major conversion is supported.
```
Comment (jorisvandenbossche) [inline docs/source/python/data.rst:576]: I think it is sufficient to say that it supports integer and float data types, I think when written generally like that it is clear this is for both signed and unsigned of all widths

File: docs/source/python/data.rst [hunk 4]
```diff
@@ -560,3 +560,59 @@ schema without having to get any of the batches.::
...
+
+   >>>  import pyarrow as pa
+   >>>  arr1 = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+   >>>  arr2 = [10, 20, 30, 40, 50, 60, 70, 80, 90]
```
Comment (jorisvandenbossche) [inline docs/source/python/data.rst:580]:
```suggestion
   >>>  arr1 = [1, 2, 3, 4, 5]
   >>>  arr2 = [10, 20, 30, 40, 50]
```

Just to keep the vertical screen estate taken by the output a bit shorter, and it is just as illustrative as the longer version I think

File: docs/source/python/data.rst [hunk 5]
```diff
@@ -560,3 +560,59 @@ schema without having to get any of the batches.::
...
+Conversion of RecordBatch do Tensor
+-----------------------------------
+
+Each array of the ``RecordBatch`` has it's own contiguous memory that is not necessarily
+adjacent to other arrays. A different memory structure that is used in machine learning
+libraries is a two dimensional array (also called a 2-dim tensor or a matrix) which takes
+only one contiguous block of memory.
+
+For this reason there is a function ``pyarrow.RecordBatch.to_tensor()`` available
+to efficiently convert tabular columnar data into a matrix.
```
Comment (jorisvandenbossche) [inline docs/source/python/data.rst:573]:
```suggestion
to efficiently convert tabular columnar data into a tensor.
```

(it's mentioned above this is also called a matrix, but then after that I would try to use as much as possible to the same word for consistency)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! Some small wording comments

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

**Pull request merged**